api/src/main/java/org/apache/iceberg/util/CharSequenceWrapper.java

```diff
@@ -29,13 +29,16 @@ public static CharSequenceWrapper wrap(CharSequence seq) {
   }
 
   private CharSequence wrapped;
+  // lazily computed & cached hashCode
+  private transient int hashCode = 0;
```

Comment:
Hmm, I'm not sure I'd set a default value of 0, since that technically can be a legit hashCode value for some series of characters no? 
I think I'd make this an Integer hashCode and have it be null to identify that nothing is cached yet.

Reply:
But good call making it transient, that's important since it shouldn't be serialized and the cache somehow used across JVMs

Reply:
I went down this rabbit hole: https://stackoverflow.com/questions/18746394/can-a-non-empty-string-have-a-hashcode-of-zero

Reply:
ah good point, thanks for finding that. Instead of making this an `Integer` I actually introduced a boolean flag. The `String' implementation does the same thing so that the hash isn't re-calculated when the hash is actually 0

Reply:
Awesome, maybe just comment in the code that we're following what java.lang.String does

Reply:
good point, I've added a comment mentioning that this follows java.lang.String